python/pyarrow/tests/test_compute.py

```diff
@@ -3277,19 +3281,106 @@ def test_expression_serialization():
         restored = pickle.loads(pickle.dumps(expr))
         assert expr.equals(restored)
 
+    if pas is not None:
+
+        test_schema = pa.schema([pa.field("i64", pa.int64()), pa.field(
+            "foo", pa.struct([pa.field("bar", pa.string())]))])
+
+        # Basic literals don't change on binding and so they will round
+        # trip without any change
+        for expr in literal_exprs:
+            serialized = expr.to_substrait(test_schema)
+            deserialized = pc.Expression.from_substrait(serialized)
+            assert expr.equals(deserialized)
+
+        # Expressions are bound when they get serialized.  Since bound
+        # expressions are not equal to their unbound variants we cannot
+        # compare the round tripped with the original
+        for expr in exprs_with_call:
+            serialized = expr.to_substrait(test_schema)
+            deserialized = pc.Expression.from_substrait(serialized)
+            # We can't compare the expressions themselves because of the bound
+            # unbound difference. But we can compare the string representation
+            assert str(deserialized) == str(expr)
+            serialized_again = deserialized.to_substrait(test_schema)
+            deserialized_again = pc.Expression.from_substrait(serialized_again)
+            assert deserialized.equals(deserialized_again)
+
+        # Expressions that have references will be normalized, on serialization,
+        # to numeric referneces
+        exprs_with_ref_norm = [pc.field(0) > 5, pc.field(0) == 5,
+                               pc.field(0) == 7,
+                               pc.field((1, 0)) == 'value',
+                               pc.field(1, 0) == 'value']
+
+        for expr, expr_norm in zip(exprs_with_ref, exprs_with_ref_norm):
+            serialized = expr.to_substrait(test_schema)
+            deserialized = pc.Expression.from_substrait(serialized)
+            assert str(deserialized) == str(expr_norm)
+            serialized_again = deserialized.to_substrait(test_schema)
+            deserialized_again = pc.Expression.from_substrait(serialized_again)
+            assert deserialized.equals(deserialized_again)
+
+        # For the special cases we get various wrinkles in serialization but we
+        # should always get the same thing from round tripping twice
+        for expr in special_cases:
+            serialized = expr.to_substrait(test_schema)
+            deserialized = pc.Expression.from_substrait(serialized)
+            serialized_again = deserialized.to_substrait(test_schema)
+            deserialized_again = pc.Expression.from_substrait(serialized_again)
+            assert deserialized.equals(deserialized_again)
+
+        # Special case, we lose the field names of struct literals
+        serialized = f.to_substrait(test_schema)
+        deserialized = pc.Expression.from_substrait(serialized)
+        assert deserialized.equals(pc.scalar({'': 1}))
```

Comment:
Should we use something else than an empty string? (eg "field0", "field1", ..) 
(although this is probably enough of a corner case to not put too much effort in it)

Reply:
Hmm, I think I'd rather leave it this way for now so it's more obvious that the fields were lost and not that they got mixed up with some other schema.  This is a bit of an obnoxious corner case.  We might want to fix this on the Substrait side.